src/tests/Interop/COM/NETClients/IDispatch/Program.cs

```diff
@@ -245,18 +245,56 @@ static void Validate_ValueCoerce_ReturnToManaged()
-            var variantException = Assert.Throws<InvalidOleVariantTypeException>(() => dispatchCoerceTesting.ReturnToManaged(0x7FFF));
-            Assert.Equal(unchecked((int)0x80131531), variantException.HResult);
-
             // Not supported source or destination type: COMException { HResult: 0x80020005 }
 
             // DISP_E_PARAMNOTFOUND: Converts to Missing
             Console.WriteLine("Converting from VT_ERROR with DISP_E_PARAMNOTFOUND should be rejected.");
             var comException = Assert.Throws<COMException>(() => dispatchCoerceTesting.ReturnToManaged(unchecked((short)((short)VarEnum.VT_ERROR | 0x8000))));
             Assert.Equal(unchecked((int)0x80020005), comException.HResult);
 
+            // Types rejected by OAVariantLib
+            VarEnum[] unsupportedTypes = 
+            {
+                VarEnum.VT_ERROR | (VarEnum)0x8000,
+            };
+
+            // Types rejected by VariantChangeTypeEx
+            VarEnum[] invalidCastTypes = 
+            {
+                VarEnum.VT_UNKNOWN,
+                VarEnum.VT_NULL,
+            };
+
+            foreach (var vt in invalidCastTypes)
+            {
+                Console.WriteLine($"Converting {vt} to int should fail from VariantChangeTypeEx.");
+                Assert.Throws<InvalidCastException>(() => dispatchCoerceTesting.ReturnToManaged((short)vt));
+            }
+
+            // Invalid: Rejected before reaching coerce
+            Console.WriteLine("Invalid variant type should throw InvalidOleVariantTypeException.");
+            var variantException = Assert.Throws<InvalidOleVariantTypeException>(() => dispatchCoerceTesting.ReturnToManaged(0x7FFF));
+            Assert.Equal(unchecked((int)0x80131531), variantException.HResult);
+
+            Console.WriteLine("Invoking void-returning method should not allocate return buffer.");
+            // E_POINTER translates to NullReferenceException
+            Assert.Throws<NullReferenceException>(() => dispatchCoerceTesting.ReturnToManaged_Void());
+
+            Console.WriteLine("Converting int to double should be supported.");
+            Assert.Equal(1234d, dispatchCoerceTesting.ReturnToManaged_Double());
+
+            Console.WriteLine("Converting int to string should be supported.");
+            Assert.Equal("1234", dispatchCoerceTesting.ReturnToManaged_String());
+
+            Console.WriteLine("Converting int to decimal should be supported.");
+            Assert.Equal(1234m, dispatchCoerceTesting.ReturnToManaged_Decimal());
+
+            Console.WriteLine("Converting int to DateTime should be supported.");
+            Assert.Equal(new DateTime(1903, 5, 18, 0, 0, 0), dispatchCoerceTesting.ReturnToManaged_DateTime());
+
+            Console.WriteLine("Converting int to System.Drawing.Color should be supported.");
+            Assert.Equal(System.Drawing.Color.FromArgb(210, 4, 0), dispatchCoerceTesting.ReturnToManaged_Color());
```

Comment:
Just returing `1234` isn't really capturing the issues here. Please at least take an input that indicates the boundaries. Min/Max and something arbitrary. The decimal and DateTime are commonly troublesome.

Reply:
I'm concerned that if we are overkill here. It's testing the behavior of `VariantChangeTypeEx` more. The response of code in this PR is just passing the correct value to `VariantChangeTypeEx`.

Reply:
I understand that concern. My perspective here is I've needed to fix interop tests/code several times because all that was tested was magic numbers like 42 or 1234 - this is lazy. Bit pattterns matter here and how we marshal the entire field of bits should be validated. Min and max values for specific data types generally do that. The fact that we are asking for the subsystem to marshal a bit pattern based on a tag is indeed the primary use case and I'm glad that is being done. I also appreciate that testing the `IDispatch` scenarios is cumbersome. All that aside, we are trying to ensure we validate existing legacy code paths and as they evolve our tests shouldn't need large updates. If we were implementing this feature new, we would test boundaries and a lot more. At a very least we should test the boundaries in this PR since we are changing very old code that many applications likely rely upon.

Reply:
We started this exercise asking the question - "What scenarios use this code path?" We are testing scenarios, **not** this code path. The fact that the `IDispatch` uses this code is an implementation detail which can be altered or even entirely replaced. If we were performing unit testing, I would agree with your analysis but we aren't. I am asking about shoring up testing of the `IDispatch` scenarios so that when the underlying logic is completely managed code we have the confidence to change it. If we leave it as such, we have "tests" that are grossly insufficient to validate much and therefore we retain the current code paths as much as possible. When we alter code paths that rely on implementation detail we should take the opportunity to ensure we have reasonable domain coverage.

The gist here is unit versus integration testing. Sadly, we do far more integration testing in CoreCLR and as such we pay a slightly different price when ensuring we have test coverage.

Reply:
Fair. If we are doing big refactor in the future, for example removing the native->managed->native step with two different mappings, we will be more confident about it.

Reply:
My suggestion here would be for the `IDispatch` API to also take a `byte[]` or `string` or `int` and use that to indicate to the other side what to return. It doesn't have to be exhaustive, just min max and arbitrary for numeric values.

Reply:
Added some tests at both native and managed side. Please help to check if the coverage is enough.